zkevm-circuits/src/blake2b/circuit/src/lib.rs

```diff
@@ -0,0 +1,1855 @@
+        // Using the formula ceil(log2(max((q + 2) * R + (n - 1) * (T + 8), 65536) + u))
+        let mut usable = (rounds + 2) * R + (inputs.len() - 1) * ( config.tail_height + 8);
+        if usable < 65536 { usable = 65536; }
+        ((usable + config.unusable_rows) as f64).log2().ceil() as u32
+    }
+
+    // Computes the array describing the optimal amounts of rows per round for the specified input parameters 
+    // of a circuit instance and all possible circuit heights. The k-th entry of this array is either the 
+    // amount of rows per round, which corresponds to a possible circuit instance with the minimum number 
+    // of columns among all circuit instances of height 2^k, which may be created for these input parameters, 
+    // or None, iff a circuit instance of height 2^k is not possible for the given parameters 
+    pub fn optimums(inputs: &[CompressionInput]) -> [Option<usize>; 32] {
+        let k = [ 
+            CompressionCircuit::<F, {ROWS_PER_ROUND[0]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[1]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[2]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[3]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[4]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[5]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[6]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[7]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[8]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[9]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[10]}>::k(inputs),
+            CompressionCircuit::<F, {ROWS_PER_ROUND[11]}>::k(inputs)];
+        
+        array::from_fn(|i| k.iter().enumerate().filter(|e| *e.1 <= i as u32).map(|e| ROWS_PER_ROUND[e.0]).max())
+    }
+}
+
+impl<F:FieldExt, const R: usize> Circuit<F> for CompressionCircuit<F,R> {
+    type Config = CompressionConfig<F,R>;
+    type FloorPlanner = SimpleFloorPlanner;
+
+    fn without_witnesses(&self) -> Self {
+        Self::default()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<F>) -> Self::Config {
+        assert!(ROWS_PER_ROUND.contains(&R), "Invalid number of rows per round!");
+        
+        let fixed = |meta: &mut ConstraintSystem<F>| meta.fixed_column();
+        let advice = |meta: &mut ConstraintSystem<F>| meta.advice_column();
+
+        let [allocated, allowed, septalookup] = array::from_fn(|_| fixed(meta));
+        let xlookup: [Column<Fixed>; 3] = array::from_fn(|_| fixed(meta));
+          
+        // Allocating the byte-column pairs required for storing the input and intermediate data 
+        // of a RoundGate class gate in R sequential rows. Storing these data requires 56 byte-column 
+        // chunks of height 8 and 32 byte-column cells  
+        let mut pairs = vec![];
+        while pairs.len() * R < 240 {
```

Comment:
nit: it might be context heavy to understand the `240` as it isn't a simple multiplication of the matrix. Consider adding a context with the required context as documentation

Reply:
As the comment says,
// Allocating the byte-column pairs required for storing the input and intermediate data 
// of a RoundGate class gate in R sequential rows. Storing these data requires 56 byte-column 
// chunks of height 8 and 32 byte-column cells "
Thus, we need 56 * 8 + 32 = 480 cells or 240 cell pairs. Should I add this to the comments?

Reply:
Probably not needed, but info that leads to the exact calculation is never enough